cmake/introspection.cmake

```diff
@@ -0,0 +1,246 @@
+  include(TestAppendRequiredLibraries)
+  test_append_socket_library(core_interface)
+endif()
+
+include(TestAppendRequiredLibraries)
+test_append_atomic_library(core_interface)
+
+check_cxx_symbol_exists(std::system "cstdlib" HAVE_STD_SYSTEM)
+check_cxx_symbol_exists(::_wsystem "stdlib.h" HAVE__WSYSTEM)
+if(HAVE_STD_SYSTEM OR HAVE__WSYSTEM)
+  set(HAVE_SYSTEM 1)
+endif()
+
+check_include_file_cxx(string.h HAVE_STRING_H)
+if(HAVE_STRING_H)
+  check_cxx_source_compiles("
+    #include <string.h>
+
+    int main()
+    {
+      char buf[100];
+      char* p{strerror_r(0, buf, sizeof buf)};
+      (void)p;
+    }
+    " STRERROR_R_CHAR_P
+  )
+endif()
+
+# Check for malloc_info (for memory statistics information in getmemoryinfo).
+check_cxx_symbol_exists(malloc_info "malloc.h" HAVE_MALLOC_INFO)
+
+# Check for mallopt(M_ARENA_MAX) (to set glibc arenas).
+check_cxx_source_compiles("
+  #include <malloc.h>
+
+  int main()
+  {
+    mallopt(M_ARENA_MAX, 1);
+  }
+  " HAVE_MALLOPT_ARENA_MAX
+)
+
+# Check for posix_fallocate().
+check_cxx_source_compiles("
+  // same as in src/util/fs_helpers.cpp
+  #ifdef __linux__
+  #ifdef _POSIX_C_SOURCE
+  #undef _POSIX_C_SOURCE
+  #endif
+  #define _POSIX_C_SOURCE 200112L
+  #endif // __linux__
+  #include <fcntl.h>
+
+  int main()
+  {
+    return posix_fallocate(0, 0, 0);
+  }
+  " HAVE_POSIX_FALLOCATE
+)
+
+# Check for strong getauxval() support in the system headers.
+check_cxx_source_compiles("
+  #include <sys/auxv.h>
+
+  int main()
+  {
+    getauxval(AT_HWCAP);
+  }
+  " HAVE_STRONG_GETAUXVAL
+)
+
+# Check for UNIX sockets.
+check_cxx_source_compiles("
+  #include <sys/socket.h>
+  #include <sys/un.h>
+
+  int main()
+  {
+    struct sockaddr_un addr;
+    addr.sun_family = AF_UNIX;
+  }
+  " HAVE_SOCKADDR_UN
+)
+
+# Check for different ways of gathering OS randomness:
+# - Linux getrandom()
+check_cxx_source_compiles("
+  #include <sys/random.h>
+
+  int main()
+  {
+    getrandom(nullptr, 32, 0);
+  }
+  " HAVE_GETRANDOM
+)
+
+# - BSD getentropy()
+check_cxx_source_compiles("
+  #include <sys/random.h>
+
+  int main()
+  {
+    getentropy(nullptr, 32);
+  }
+  " HAVE_GETENTROPY_RAND
+)
+
+
+# - BSD sysctl()
+check_cxx_source_compiles("
+  #include <sys/types.h>
+  #include <sys/sysctl.h>
+
+  #ifdef __linux__
+  #error Don't use sysctl on Linux, it's deprecated even when it works
+  #endif
+
+  int main()
+  {
+    sysctl(nullptr, 2, nullptr, nullptr, nullptr, 0);
+  }
+  " HAVE_SYSCTL
+)
+
+# - BSD sysctl(KERN_ARND)
+check_cxx_source_compiles("
+  #include <sys/types.h>
+  #include <sys/sysctl.h>
+
+  #ifdef __linux__
+  #error Don't use sysctl on Linux, it's deprecated even when it works
+  #endif
+
+  int main()
+  {
+    static int name[2] = {CTL_KERN, KERN_ARND};
+    sysctl(name, 2, nullptr, nullptr, nullptr, 0);
+  }
+  " HAVE_SYSCTL_ARND
+)
+
+check_cxx_source_compiles("
+  int foo(void) __attribute__((visibility(\"default\")));
+  int main(){}
+  " HAVE_DEFAULT_VISIBILITY_ATTRIBUTE
```

Comment:
Shouldn't this be stopping the build if this check fails and `REDUCE_EXPORTS` is used (same as master)? Although, given they are unused, could be better to just remove entirely, rather than porting and not matching the behaviour. I'll open a PR.

Reply:
See #30590.